src/index/segment_reader.rs

```diff
@@ -213,83 +213,137 @@ impl SegmentReader {
         segment: &Segment,
         custom_bitset: Option<AliveBitSet>,
     ) -> crate::Result<SegmentReader> {
-        let start = std::time::Instant::now();
-        let segment_ref = SegmentRef::new(segment);
-        let termdict_composite_fut = executor.spawn(move || {
-            let termdict_file = segment_ref.open_read(SegmentComponent::Terms)?;
-            Ok(CompositeFile::open(&termdict_file)?)
-        })?;
-
-        let segment_ref = SegmentRef::new(segment);
-        let store_file_fut =
-            executor.spawn(move || Ok(segment_ref.open_read(SegmentComponent::Store)?))?;
-
-        crate::fail_point!("SegmentReader::open#middle");
+        use std::sync::Mutex;
 
-        let segment_ref = SegmentRef::new(segment);
-        let postings_composite_fut = executor.spawn(move || {
-            let postings_file = segment_ref.open_read(SegmentComponent::Postings)?;
-            Ok(CompositeFile::open(&postings_file)?)
-        })?;
+        let schema = segment.schema();
 
-        let segment_ref = SegmentRef::new(segment);
-        let positions_composite_fut = executor.spawn(move || {
-            if let Ok(positions_file) = segment_ref.open_read(SegmentComponent::Positions) {
-                Ok(CompositeFile::open(&positions_file)?)
-            } else {
-                Ok(CompositeFile::empty())
-            }
-        })?;
+        // Helper macro to spawn a task that stores result in a Mutex
+        macro_rules! spawn_load {
+            ($scope:expr, $result:expr, $body:expr) => {
+                $scope.spawn(|_| {
+                    *$result.lock().unwrap() = Some($body);
+                });
+            };
+        }
 
-        let schema = segment.schema();
+        // Use scope-based parallelism to avoid deadlock when called from within executor.map()
+        executor.install_and_scope(|| {
+            // Check if we have a thread pool available
+            if let Some(pool) = executor.thread_pool() {
```

Comment:
nit: I might avoid one level of nesting by short-circuit returning in the other case. E.g.
```
let pool = match executor.thread_pool() {
  None => { return SegmentReader::open(...); },
  Some(pool) => pool
};
...
```